packageurl.go

```diff
@@ -331,6 +354,43 @@ func typeAdjustName(purlType, name string) string {
 	return name
 }
 
+// validQualifierKey validates a qualifierKey against our QualifierKeyPattern.
 func validQualifierKey(key string) bool {
 	return QualifierKeyPattern.MatchString(key)
 }
+
+// validCustomRules evaluates additional rules for each package url type, as specified in the package-url specification.
+// On success, it returns nil. On failure, a descriptive error will be returned.
+func validCustomRules(purlType, name, ns, version string, qualifiers Qualifiers) error {
+	q := qualifiers.Map()
+	switch purlType {
+	case TypeConan:
+		if ns != "" {
+			if val, ok := q["channel"]; ok {
+				if val == "" {
+					return errors.New("the qualifier channel must be not empty if namespace is present")
+				}
+			} else {
+				return errors.New("channel qualifier does not exist")
+			}
+		} else {
+			if val, ok := q["channel"]; ok {
+				if val != "" {
+					return errors.New("namespace is required if channel is non empty")
+				}
+			}
+		}
+	case TypeSwift:
+		if ns == "" {
+			return errors.New("namespace is required")
+		}
+		if version == "" {
+			return errors.New("version is required")
+		}
+	case TypeCran:
+		if version == "" {
+			return errors.New("version is required")
```

Comment:
Why would you always require a version here and above for Swift?

Reply:
Because otherwise the test cases fail. I wasn't sure if the test cases are wrong or the implementation. I thought it must be the latter, hence I added this additional version check for the swift package url.

Reply:
The test case are likely incorrect then. It is always OK to have a Package URL without a version. It may not be super useful but this is not invalid adn has its use: for instance, a dependency or vulnerability to a package that would be further specified as version range would not have a version, but possible a version range as a qualifier.
See the WIP at https://github.com/package-url/purl-spec/pull/139

Let me review the tests as this is not right!

Reply:
If I remove the version comparisons, I will get these errors here while running `make test`:
```
=== RUN   TestFromStringExamples
    packageurl_test.go:174: invalid cran purl without version did not fail and returned packageurl.PackageURL{Type:"cran", Namespace:"", Name:"A3", Version:"", Qualifiers:packageurl.Qualifiers{}, Subpath:""}
    packageurl_test.go:174: invalid swift purl without version did not fail and returned packageurl.PackageURL{Type:"swift", Namespace:"github.com/Alamofire", Name:"Alamofire", Version:"", Qualifiers:packageurl.Qualifiers{}, Subpath:""}

```

Reply:
I do agree that the test case likely needs some updating. _However_ that can be done outside of this PR IMHO.

Reply:
I suggest, I set the missing symlink in this PR for the license file and then we are getting this merged, because we need this functionality in one of our tools at sigstore.

If merged, we can have a closer look on the test cases and verify if we *really* require the version for a few package urls.